src/Core/Framework/DataAbstractionLayer/Dbal/EntityReader.php

```diff
@@ -807,7 +812,9 @@ private function loadManyToManyWithCriteria(
         }
         unset($row);
 
-        $fieldCriteria->setIds($ids);
+        if (\count($ids) !== 0) {
+            $fieldCriteria->setIds($ids);
+        }
```

Comment:
In this case when `$ids` result is empty, property `ids` of `$fieldCriteria` (Criteria) is still an empty array (its default value) and thus all entities are loaded